contracts/modules/Experimental/Checkpoints/PLCRVotingCheckpoint.sol

```diff
@@ -0,0 +1,314 @@
+        uint256 _revealDuration,
+        uint256 _totalProposals,
+        uint256 _proposedQuorum,
+        uint256 _checkpointId,
+        uint256 _startTime
+    )
+        internal
+    {
+        // Sanity checks
+        _validValueCheck(_commitDuration);
+        _validValueCheck(_revealDuration);
+        _validValueCheck(_proposedQuorum);
+        // Overflow check
+        require(
+            uint64(_commitDuration) == _commitDuration && 
+            uint64(_revealDuration) == _revealDuration &&
+            uint64(_startTime) == _startTime &&
+            uint24(_totalProposals) == _totalProposals,
+            "Parameter values get overflowed"
+        );
+        require(_startTime >= now, "Invalid start time");
+        require(_totalProposals > 1, "Invalid number of totalProposals");
+        uint256 _ballotId = ballots.length;
+        ballots.push(Ballot(
+            _proposedQuorum,
+            _checkpointId,
+            uint64(_commitDuration),
+            uint64(_revealDuration),
+            uint64(_startTime),
+            uint24(_totalProposals),
+            uint32(0),
+            true
+        ));
+        emit BallotCreated(_ballotId, _commitDuration, _revealDuration, _totalProposals, _startTime, _proposedQuorum, _checkpointId);
+    }
+
+    /**
+     * @notice Used to commit the vote
+     * @param _ballotId Given ballot Id
+     * @param _secretVote It is secret hash value (hashed offchain)
+     */
+    function commitVote(uint256 _ballotId, bytes32 _secretVote) external {
+        _validBallotId(_ballotId);
+        require(getCurrentBallotStage(_ballotId) == Stage.COMMIT, "Not in commit stage");
+        Ballot storage ballot = ballots[_ballotId];
+        require(ballot.proposalToVote[msg.sender].secretVote == bytes32(0), "Already voted");
+        require(_secretVote != bytes32(0), "Invalid vote");
+        require(ballot.isActive, "Inactive ballot");
+        uint256 weight = ISecurityToken(securityToken).balanceOfAt(msg.sender, ballot.checkpointId);
+        require(weight > 0, "Zero weight is not allowed");
+        ballot.proposalToVote[msg.sender] = Vote(0, _secretVote);
+        emit VoteCommited(msg.sender, weight, _ballotId, _secretVote);
+    }
+
+    /**
+     * @notice Used to reveal the vote
+     * @param _ballotId Given ballot Id
+     * @param _choiceOfProposal Proposal chossed by the voter.
+     * @param _salt used salt for hashing (unique for each user)
+     */
+    function revealVote(uint256 _ballotId, uint256 _choiceOfProposal, uint256 _salt) external {
+        _validBallotId(_ballotId);
+        require(getCurrentBallotStage(_ballotId) == Stage.REVEAL, "Not in reveal stage");
+        Ballot storage ballot = ballots[_ballotId];
+        require(ballot.isActive, "Inactive ballot");
+        require(ballot.proposalToVote[msg.sender].secretVote != bytes32(0), "Should be commit vote first");
+        require(_choiceOfProposal < ballot.totalProposals, "Invalid proposal choice");
```

Comment:
If proposals start with id 1 then we should use `<=` instead of `<` here. If proposals start with id 0 (as they should), we should make it clear here. While creating a proposal, it takes the number of proposals as input and minimum is set to two. That suggests that valid proposals are 1 and 2 rather than 0 and 1.

Reply:
I think only need to add the comment to clear it. As we are starting the proposal Id with 0 here in this module